### PR TITLE
FEATURE: Add ParsePartials cache to the fusion parser

### PR DESCRIPTION
**Pull request by mhsdesign**
Resolves: #3661

The New Fusion Parser now uses a `Neos_Fusion_ParsePartials` cache to store the generated FileAst _*_ for each Fusion File. Those File partials are flushed via file monitor when the source Fusion file was changed.
The generated merged array Fusion tree for a Fusion DSL is also cached seperately in this cache.

With this cache layer in place, only changed files are reparsed which **speeds up Fusion parsing** in Development mode by **factor of 2 to 8** depending on context. _**_

The ParsePartials cache is disabled in Production mode as this one already uses the ObjectTreeCache.

since we cache mostly AstNode objects its recommended to have the php extension `igbinary` installed.

the cache can be turned of by disabling: `Neos.Fusion.enableParsePartialsCache`

_*_ there are now more representations of the Fusion code - we created a step in between where we create AstNode Objects
see for more infos the PR of the New Fusion Parser: https://github.com/neos/neos-development-collection/pull/3497

_**_ in my tests, when the cached is warmed _***_, performance boost might not be notable on small projects.

_***_ of course is the cache warmed - as always for the statistics ;)

### CODE REVIEW

**Comment by mhsdesign** [comment]
tests fail because we need to mock the parser cache helper ...

**Comment by mficzel** [comment]
Can we rebase this on master ... would love to add this still

**Comment by mhsdesign** [comment]
nope im on it ;)

File: Neos.Fusion/Classes/Core/Cache/FusionParserCache.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
+<?php
```
Comment (mhsdesign) [inline Neos.Fusion/Classes/Core/Cache/FusionParserCache.php:1]: add strict types ...

**Comment by mficzel** [comment]
@mhsdesign naming is generally fine only thing that looks a bit weird is that the dsl and the file cache are handled slightly differently. 

Why not make cacheByIdentifier private and have a function `cacheForDsl(string $code, \Closure $closure)` alongside `cacheForFilePath(string $path, \Closure $closure)`.

That way those methods would create the cache identifier with prefix that and then pass the caching to the protected cacheByIdentifier method.

**Comment by mficzel** [comment]
Also would'nt it be possible to add the functionality to the FusionCachingAspect that is already there. I know i said i dislike AOP for caching but that is already introduced. 

Advantage of AOP would be in this case that the code would not be active once the cache is disabled so it dies not even have to check on runtime wether it is enabled

**Comment by mhsdesign** [comment]
i get the argument, of not even checking for the setting ... but wouldnt this conflict with: https://github.com/neos/neos-development-collection/issues/3216 either way in long term?

**Comment by mficzel** [comment]
@mhsdesign i agree, was not aware of #3216 and considered it more like a smell. Then let us proceed with the path you took. The comments before are still valid i think.

**Comment by mficzel** [comment]
Added 'file' and 'dsl' prefixes for the cache identifiers.

**Review by mficzel — APPROVED** [review]
I like speed

File: Neos.Fusion/Classes/Core/Cache/ParserCache.php [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        if (str_contains($fusionFileName, '://')) {
+            $fusionFileName = $this->getAbsolutePathForPackageRessourceUri($fusionFileName);
+        }
+
+        $realPath = realpath($fusionFileName);
+        if ($realPath === false) {
+            throw new \InvalidArgumentException("Couldn't resolve realpath for: '$fusionFileName'");
+        }
+
+        $flowRoot = defined('FLOW_PATH_ROOT') ? FLOW_PATH_ROOT : '';
```
Comment (kdambekalns) [inline Neos.Fusion/Classes/Core/Cache/ParserCache.php:116]: That should always be defined… why did you add that check?

File: Neos.Fusion/Classes/Core/Cache/ParserCache.php [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        $realFusionFilePathWithoutRoot = str_replace($flowRoot, '', $realPath);
+        return 'file_' . md5($realFusionFilePathWithoutRoot);
+    }
+
+    /**
+     * Uses the same technic to resolve a package resource uri like flow.
```
Comment (kdambekalns) [inline Neos.Fusion/Classes/Core/Cache/ParserCache.php:122]:
```suggestion
     * Uses the same technique to resolve a package resource URI like Flow.
```

**Review by kdambekalns — COMMENTED** [review]
Looking good, will do some testing…

File: Neos.Fusion/Classes/Core/Cache/ParserCache.php (hunk 3 above)
Comment (mhsdesign) [reply Neos.Fusion/Classes/Core/Cache/ParserCache.php:122]: quite eloquent ^^

**Review by kdambekalns — APPROVED** [review]
Ok, as far as I can tell this works fine.

**Comment by mhsdesign** [comment]
thx for testing :D

**Comment by mhsdesign** [comment]
i would like to see if one really `feels` the difference on larger projects.
(sure the speed increase on the neos demo sounds much - but due to the fact that a lot will happen when viewing a site - last but not least rendering the fusion - the speed increase might not be that notable for small projects)

**Comment by mhsdesign** [comment]
TAKEN FROM HEAD

----

# Make the new Fusion Parser faster than the old! _*_

_*_ in my tests, when the cached is warmed.

| ProjectSize | NewParserWithWarmedCache (ms) | OldParser (ms) | NewParser (ms) |
|------------|--------------------|-----------------|------------------|
| NeosDemo |  7  | 15 |  30   |
| NeosDemo + includes  |  31  | 42 |  76   |
| 6x _***_ NeosDemo + includes |  31  | 270 | 447   |
| larger project | 16  | 65  | 123    |
| larger project + includes | 40  | 97  | 189    |

_*** all fusion files were 6 times repeatedly in onefile included and then parsed._


### parser performace related development *comments* (nearly articles):
[multiple regex are slower than combining them](https://github.com/neos/neos-development-collection/pull/3497#issuecomment-964602190)
[tried out parsica, and figured and parsing time is not linear](https://github.com/neos/neos-development-collection/pull/3497#issuecomment-974914171)
[partial merged array tree cache v1 - speed up by 2x](https://github.com/neos/neos-development-collection/pull/3497#issuecomment-1025183821)


test command for testing this:
```php
class FusionCommandController extends CommandController
{
    public function timeCommand()
    {
        $fusionCode = "
include: resource://Neos.Fusion/Private/Fusion/Root.fusion
include: resource://Neos.Fusion.Form/Private/Fusion/Root.fusion
include: resource://Neos.Neos/Private/Fusion/Root.fusion
include: resource://Neos.NodeTypes.ContentReferences/Private/Fusion/Root.fusion
include: resource://Neos.NodeTypes.AssetList/Private/Fusion/Root.fusion
include: resource://Neos.NodeTypes.Form/Private/Fusion/Root.fusion
include: resource://Neos.NodeTypes.ColumnLayouts/Private/Fusion/Root.fusion
include: resource://Neos.NodeTypes.Html/Private/Fusion/Root.fusion
include: resource://Neos.NodeTypes.Navigation/Private/Fusion/Root.fusion
include: resource://Neos.NodeTypes/Private/Fusion/Root.fusion
include: resource://Neos.Neos.Ui/Private/Fusion/Root.fusion
include: resource://Neos.Seo/Private/Fusion/Root.fusion
include: resource://Flowpack.Neos.FrontendLogin/Private/Fusion/Root.fusion
include: **/*.fusion";

        $resolveForRelativePaths = "resource://Neos.Demo/Private/Fusion/Root.fusion";

        // flow cache:flushone --identifier Neos_Fusion_ParsePartials
        $start = hrtime(true);
        // for ($x = 0; $x <= 6; $x++) {
        $tree = (new Parser())->parse($fusionCode, $resolveForRelativePaths);
        // }

        // feel free to download the old Parser and put it somewhere,
        // if you do this, make sure you rename the old parser, remove the interface, and also replace recursive instantiation with `new self()` instead of `new Parser`
//         $old = (new ParserOld())->parse($fusionCode, $resolveForRelativePaths);
//         Assert::assertSame($old, $tree);
        $end = hrtime(true);
        echo ($end - $start) / 1000000; // get Milliseconds
    }
}
```

**Comment by mhsdesign** [comment]
im just thinking if the switch to enable the cache should be also `setable` on the parser like: `$parser->enableParsePartialCache(false)` 
would be usefull if you want to use the parser for own stuff and dont depend on yaml config...
